filelog.go

```diff
@@ -345,6 +357,117 @@ func (w *FileLogWriter) archiveFiles(dir string) error {
+		return
+	}
+
+	// Defer closing the underlying file
+	defer func() {
+		err = compressedFile.Close()
+		if err != nil {
+			fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't close file %q: %s\n", w.filename, compressedInprogressFilename, err)
+		}
+	}()
+
+	var compressedFileWriter io.Writer
+	switch method {
+	case "gz":
+		compressedFileWriter = gzip.NewWriter(compressedFile)
+	case "zip":
+		// zipWriter is the outer container, compressedFileWriter is an entry inside the zip
+		zipWriter := zip.NewWriter(compressedFile)
+
+		// Defer closing the zip writer
+		defer func() {
+			err = zipWriter.Close()
+			if err != nil {
+				fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't close zip writer on %q: %s\n", w.filename, compressedInprogressFilename, err)
+			}
+		}()
+
+		compressedFileWriter, err = zipWriter.Create(basename)
+		if err != nil {
+			fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't open zip file entry: %s\n", w.filename, err)
+			return
+		}
+	default:
+		fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Unknown compression method: %q\n", w.filename, method)
+		return
+	}
+
+	// Defer closing the writer, but only if we have a io.WriteCloser
+	defer func() {
+		// gzip compression requires closure, but zip doesn't have a Close() method
+		writeCloser, ok := compressedFileWriter.(io.WriteCloser)
+		if ok {
+			err = writeCloser.Close()
+			if err != nil {
+				fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't finalize compression on %q: %s\n", w.filename, compressedInprogressFilename, err)
+			}
+		}
+	}()
+
+	// If we return early after this point, we need to cleanup the .inprogress file
+	cleanupInprogress := true
+	defer func() {
+		if cleanupInprogress {
+			err = os.Remove(compressedInprogressFilename)
+			if err != nil {
+				fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't remove temporary file %q: %s\n", w.filename, err)
+			}
+		}
+	}()
+
+	// Read plain file, write to compressed file
+	_, err = io.Copy(compressedFileWriter, plainFile)
+	if err != nil {
+		fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't write compressed file %q: %s\n", w.filename, compressedInprogressFilename, err)
+		return
+	}
+
+	// Since all we do is read from plainFile, we can ignore any error closing it
+	_ = plainFile.Close()
+
+	// Rename compressed file
+	err = os.Rename(compressedInprogressFilename, compressedFilename)
+	if err != nil {
+		fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't rename file %q to %q: %s\n", w.filename, compressedInprogressFilename, compressedFilename, err)
+		return
+	}
+
+	// Upon successfully renaming the file, we no longer need to cleanup the .inprogress file
+	cleanupInprogress = false
+
+	// Delete plain file
+	err = os.Remove(filename)
+	if err != nil {
+		fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't remove file %q: %s\n", w.filename, filename, err)
+
+		// If we can't remove the old plain file, delete the compressed file so we don't affect the
```

Comment:
This kind of gets back to the complexity of depending on lexographic sorting for retention - ideally I think we'd:
- iterate over the files in the directory and filter by prefix
- parse the dates and extensions into structs
- purge any files where the date is beyond the retention period
- compress any files where the extension isn't right

It seems like a more robust system - we could ignore files that don't have the expected format (files that are rolled more than once a day), we could enforce compression even if we got stopped while compressing a file, we could clean up inprogress files if they got left behind.